pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/v2/builder/StarTreeIndexSeparator.java

```diff
@@ -0,0 +1,154 @@
+    _indexFileChannel = new RandomAccessFile(indexFile, "r").getChannel();
+    _builderConfigList = extractBuilderConfigs(metadataProperties);
+    _totalDocsList = extractTotalDocsList(metadataProperties);
+  }
+
+  private List<Map<StarTreeIndexMapUtils.IndexKey, StarTreeIndexMapUtils.IndexValue>> extractIndexMap(File indexMapFile,
+      int numStarTrees) {
+    try (InputStream inputStream = new FileInputStream(indexMapFile)) {
+      return StarTreeIndexMapUtils.loadFromInputStream(inputStream, numStarTrees);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public List<Integer> extractTotalDocsList(PropertiesConfiguration metadataProperties) {
+    List<Integer> totalDocsList = new ArrayList<>(_indexMapList.size());
+    for (int i = 0; i < _indexMapList.size(); i++) {
+      Configuration metadata = metadataProperties.subset(StarTreeV2Constants.MetadataKey.getStarTreePrefix(i));
+      totalDocsList.add(i, metadata.getInt(StarTreeV2Constants.MetadataKey.TOTAL_DOCS));
+    }
+    return totalDocsList;
+  }
+
+  /**
+   * Extract the list of {@link StarTreeV2BuilderConfig} for each of the star-tree present in the given metadata
+   * properties.
+   * @param metadataProperties index metadata properties
+   * @return List of {@link StarTreeV2BuilderConfig}
+   */
+  public List<StarTreeV2BuilderConfig> extractBuilderConfigs(PropertiesConfiguration metadataProperties) {
+    List<StarTreeV2BuilderConfig> builderConfigList = new ArrayList<>(_indexMapList.size());
+    for (int i = 0; i < _indexMapList.size(); i++) {
+      Configuration metadata = metadataProperties.subset(StarTreeV2Constants.MetadataKey.getStarTreePrefix(i));
+      builderConfigList.add(i, StarTreeV2BuilderConfig.fromIndexConfig(new StarTreeIndexConfig(
+          Lists.newArrayList(metadata.getStringArray(StarTreeV2Constants.MetadataKey.DIMENSIONS_SPLIT_ORDER)),
+          Lists.newArrayList(
+              metadata.getStringArray(StarTreeV2Constants.MetadataKey.SKIP_STAR_NODE_CREATION_FOR_DIMENSIONS)),
+          Lists.newArrayList(metadata.getStringArray(StarTreeV2Constants.MetadataKey.FUNCTION_COLUMN_PAIRS)),
+          metadata.getInt(StarTreeV2Constants.MetadataKey.MAX_LEAF_RECORDS))));
+    }
+    return builderConfigList;
+  }
+
+  /**
+   * Extract star-tree index files of the star-tree represented by the builderConfig.
+   * The extracted star-tree index files are written to the provided starTreeOutputDir.
+   *
+   * @param starTreeOutputDir output directory for extracted star-trees
+   * @param builderConfig {@link StarTreeV2BuilderConfig} of the star-tree to separate
+   * @return if star-tree exist then total docs in the separated tree, else -1
+   * @throws IOException
+   */
+  public int separate(File starTreeOutputDir, StarTreeV2BuilderConfig builderConfig)
+      throws IOException {
+    int treeIndex = _builderConfigList.indexOf(builderConfig);
+    if (treeIndex == -1) {
+      return treeIndex;
```

Comment:
(minor) this should be `return -1` (although the same effect, but logically it should return the docs)